src/components/tl/ucp/reduce/reduce_srg_knomial.c

```diff
@@ -0,0 +1,334 @@
+                                        ucc_schedule_t **frag_p)
+{
+    ucc_tl_ucp_team_t   *tl_team  = ucc_derived_of(team, ucc_tl_ucp_team_t);
+    ucc_base_coll_args_t args     = *coll_args;
+    ucc_mrange_uint_t   *p        = &tl_team->cfg.reduce_srg_kn_radix;
+    ucc_rank_t           trank    = UCC_TL_TEAM_RANK(tl_team);
+    ucc_schedule_t      *schedule;
+    ucc_coll_task_t     *task, *rs_task;
+    ucc_status_t         status;
+    ucc_kn_radix_t       radix, cfg_radix;
+    size_t               count;
+    ucc_datatype_t       dt;
+    void                 *rs_rbuf, *rs_sbuf;
+    ucc_tl_ucp_schedule_t *rsg_schedule;
+    ucc_memory_type_t mt;
+
+    rsg_schedule = ucc_derived_of(sp, ucc_tl_ucp_schedule_t);
+    status = ucc_tl_ucp_get_schedule(tl_team, coll_args,
+                                     (ucc_tl_ucp_schedule_t **)&schedule);
+    if (ucc_unlikely(UCC_OK != status)) {
+        return status;
+    }
+
+    if (UCC_IS_ROOT(coll_args->args, trank)) {
+        dt      = coll_args->args.dst.info.datatype;
+        mt      = coll_args->args.dst.info.mem_type;
+        if (UCC_IS_INPLACE(coll_args->args)) {
+            rs_rbuf = rsg_schedule->scratch_mc_header->addr;
+            rs_sbuf = coll_args->args.dst.info.buffer;
+        } else {
+            rs_rbuf = coll_args->args.dst.info.buffer;
+            rs_sbuf = coll_args->args.src.info.buffer;
+        }
+        count = coll_args->args.dst.info.count;
+    } else {
+        dt      = coll_args->args.src.info.datatype;
+        mt      = coll_args->args.src.info.mem_type;
+        rs_rbuf = rsg_schedule->scratch_mc_header->addr;
+        rs_sbuf = coll_args->args.src.info.buffer;
+        count   = coll_args->args.src.info.count;
+    }
+
+    if (coll_args->mask & UCC_BASE_CARGS_MAX_FRAG_COUNT) {
+        count = coll_args->max_frag_count;
+    }
+
+    args.args.flags             &= ~UCC_COLL_ARGS_FLAG_IN_PLACE;
+    args.args.dst.info.buffer   = rs_rbuf;
+    args.args.dst.info.count    = count;
+    args.args.dst.info.datatype = dt;
+    args.args.dst.info.mem_type = mt;
+    args.args.src.info.buffer   = rs_sbuf;
+    args.args.src.info.count    = count;
+    args.args.src.info.datatype = dt;
+    args.args.src.info.mem_type = mt;
+
+    cfg_radix = ucc_tl_ucp_get_radix_from_range(tl_team,
+                                                count * ucc_dt_size(dt),
+                                                mt, p, 4);
+    radix     = ucc_knomial_pattern_get_min_radix(cfg_radix,
+                                                  UCC_TL_TEAM_SIZE(tl_team),
+                                                  count);
+
+    /* 1st step of reduce: knomial reduce_scatter */
+    UCC_CHECK_GOTO(ucc_tl_ucp_reduce_scatter_knomial_init_r(&args, team, &task,
+                                                            radix),
+                   out, status);
+    UCC_CHECK_GOTO(ucc_schedule_add_task(schedule, task), out, status);
```

Comment:
I think there is memory leak here in case of error (e.g. the scratch buffer, or the schedule handle)

Reply:
I don't see memleak, if init frag return error then ucc_schedule_pipelined_init also returns error and then err_free_scratch label has all necessary cleanup